README.md
# Fn

Functional utility library for PHP.


## API

### partial


### compose


### identity


### constant


### always


### never


### cat


### moreThanOrEqualTo


### lessthan


### equalTo


### notEqualTo


### even


### odd


### both


### invoker


### invokeAll


### invoke


### negate


### conditional


### doWhen


### errorThrower


### conditionalThrower


### throwWhen


### throwError


### resolve


### resolver


### whenSetInvoker


### keySetChecker


### isKeySet


### accessor


### cb


### zip


### mapAssoc


### pick


### pluck

Plucks a property from a collection of associative arrays.

eg:

```php
	Fn\pluck([
			['name' => 'moe', 'age' => 45],
			['name' => 'larry', 'age' => 55],
			['name' => 'curly', 'age' => 65]
		], 'name');
  // --> ['moe', 'larry', 'curly']
```

### concat

Add a value to an array, or merge a set of values.

```php
    Fn\concat(['a', 'b'], 'c');  				  // ['a', 'b', 'c']
    Fn\concat(['a', 'b'], ['c', 'd']);  	// ['a', 'b', 'c', 'd']
```

### any

Returns true if any item in the array passes a predicate

#### Example:

```php
Fn\any([1, 3, 4, 5, 9], Fn\even()) // true
Fn\any([1, 3, 5, 9], Fn\even()) // false
```

### all

Returns true if every item in the array passes a predicate

#### Example:

```php
Fn\all([1, 3, 5, 9], Fn\odd()) // true
Fn\all([1, 3, 4, 5, 9], Fn\even()) // false
```

### joinOr


### joinAnd


### factoryMap


### factory


### caller


### countWhere


### reduceAssoc


### times

Invoke a callable some number of times.

```php
Fn\times(3, function($i) {
	echo "call #$i";
})

// call 1
// call 2
// call 3
```

CHANGELOG.md
# v1.3.0

* ADD: `Fn\times()`

# v1.2.0

* ADD: `Fn\find()`

# v1.1.0

* ADD: `Fn\pluck()`, `Fn\concat()`

# v1.0.0

Initial Release.


tests/FnTest.php
<?php

use Aeris\Fn;
use Aeris\FnTest\Fixture\SomeClass;
use Aeris\Spy;

class FnTest extends \PHPUnit_Framework_TestCase {

	/** @test */
	public function method_any() {
		$this->assertTrue(Fn\any([false, true, false]));
		$this->assertFalse(Fn\any([false, false, false]));

		$isShazaam = function ($arg) {
			return $arg === 'shazaam';
		};
		$this->assertTrue(Fn\any(['baz', 'shazaam', 'baz'], $isShazaam));
		$this->assertFalse(Fn\any(['baz', 'baz', 'baz'], $isShazaam));
		$this->assertFalse(Fn\any([]), $isShazaam);
	}

	/** @test */
	public function method_factory() {
		$someClassFactory = Fn\factory('Aeris\FnTest\Fixture\SomeClass');

		$someClassInstance = $someClassFactory('a', 'b', 'c');

		$this->assertInstanceOf('Aeris\FnTest\Fixture\SomeClass', $someClassInstance);
		$this->assertEquals(['a', 'b', 'c'], $someClassInstance->ctorArgs);
	}

	/** @test */
	public function method_factoryMap() {
		$someClassMapper = Fn\factoryMap('Aeris\FnTest\Fixture\SomeClass');

		$someClassInstances = $someClassMapper(['a', 'b', 'c']);

		$this->assertEquals(count($someClassInstances), 3);
		$this->assertInstanceOf('Aeris\FnTest\Fixture\SomeClass', $someClassInstances[0]);
		$this->assertInstanceOf('Aeris\FnTest\Fixture\SomeClass', $someClassInstances[1]);
		$this->assertInstanceOf('Aeris\FnTest\Fixture\SomeClass', $someClassInstances[2]);
		$this->assertEquals(['a'], $someClassInstances[0]->ctorArgs);
		$this->assertEquals(['b'], $someClassInstances[1]->ctorArgs);
		$this->assertEquals(['c'], $someClassInstances[2]->ctorArgs);
	}

	/** @test */
	public function method_find() {
		$oneThroughTen = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10];
		$moreThanFive = Spy::returnsUsing(function ($n) {
			return $n > 5;
		});

		$this->assertEquals(6, Fn\find($oneThroughTen, $moreThanFive), 'Should return the first value to pass the predicate');

		// Should only call predicate until value is found (for performance)
		$moreThanFive->shouldHaveBeenCalled()
			->times(6);

		$this->assertNull(Fn\find([1, 2, 3], $moreThanFive), 'Should return null if none match.');

		$this->assertNull(Fn\find([], Fn\always()), 'Should return null if collection is empty');

		$spy = Spy::returns(true);
		$this->assertEquals('bar', Fn\find(['foo' => 'bar'], $spy), 'Should work with associateive arrays');

		// Should call spy with ($val, $key);
		$spy->shouldHaveBeenCalled()
			->with('bar', 'foo');
	}


	/** @test */
	public function method_caller() {
		// Basic example, though result is useless
		$passThruCaller = Fn\caller('passThru');
		$this->assertEquals('foo', $passThruCaller(new SomeClass(), 'foo'));

		// A more useful example, using mapping.
		$objects = [
			new SomeClass(),
			new SomeClass(),
			new SomeClass()
		];
		$valList = array_map(Fn\caller('getSomeProp'), $objects);
		$this->assertEquals(['someVal', 'someVal', 'someVal'], $valList);
	}

	/** @test */
	public function method_countWhere() {
		$this->assertEquals(3, Fn\countWhere([
			'foo',
			'bar',
			'foo'
		], true), 'Should return array count, when predicate is true');
		$this->assertEquals(0, Fn\countWhere(['foo', 'bar', 'foo'], false), 'Should return 0, when predicate is false');

		$this->assertEquals(2, Fn\countWhere(['foo', 'bar', 'foo'], function ($item) {
			return $item == 'foo';
		}), 'Should return the count of items passing a predicate fn.');
	}


	/** @test */
	public function method_accessor() {
		$arr = ['foo' => 'bar'];
		$fooAccessor = Fn\accessor('foo');

		$this->assertEquals('bar', $fooAccessor($arr));
	}

	/** @test */
	public function method_all() {
		$arr = [3, 4, 5, 6];

		$this->assertTrue(Fn\all($arr, Fn\moreThanOrEqualTo(3)));
		$this->assertFalse(Fn\all($arr, Fn\even()));
		$this->assertFalse(Fn\all($arr, Fn\odd()));
	}

	/** @test */
	public function method_even() {
		$isEven = Fn\even();
		$this->assertTrue($isEven(4));
		$this->assertFalse($isEven(15));
	}

	/** @test */
	public function method_odd() {
		$isOdd = Fn\odd();
		$this->assertFalse($isOdd(4));
		$this->assertTrue($isOdd(15));
	}

	/** @test */
	public function method_always() {
		$always = Fn\always();
		$this->assertTrue($always());
		$this->assertTrue($always(true));
		$this->assertTrue($always(false));
	}

	/** @test */
	public function method_never() {
		$never = Fn\never();
		$this->assertFalse($never());
		$this->assertFalse($never(true));
		$this->assertFalse($never(false));
	}

	/** @test */
	public function method_resolver() {
		$resolve = Fn\resolver('a', 'b', 'c');

		$this->assertTrue($resolve(true));
		$this->assertFalse($resolve(false));

		$this->assertTrue($resolve(function ($a, $b, $c) {
			return $a == 'a' && $b == 'b' && $c == 'c';
		}));
		$this->assertFalse($resolve(function ($a, $b, $c) {
			return false;
		}));
	}

	/** @test */
	public function method_joinAnd() {
		$predicates = [
			Fn\even(),
			Fn\moreThanOrEqualTo(10),
		];
		$checkAll = Fn\joinAnd($predicates);

		$this->assertTrue($checkAll(12));
		$this->assertTrue($checkAll(14));
		$this->assertFalse($checkAll(8));
		$this->assertFalse($checkAll(13));
	}

	/** @test */
	public function method_joinOr() {
		$predicates = [
			Fn\even(),
			Fn\moreThanOrEqualTo(10),
		];
		$checkAny = Fn\joinOr($predicates);

		$this->assertTrue($checkAny(2));
		$this->assertTrue($checkAny(11));
		$this->assertFalse($checkAny(7));
	}

	/** @test */
	public function method_reduceAssoc() {
		$arr = [
			'foo' => 'bar',
			'faz' => 'baz'
		];

		$reduced = Fn\reduceAssoc($arr, function ($reduced, $val, $key) {
			return $reduced . $key . $val;
		}, '');

		$this->assertEquals('foobarfazbaz', $reduced);
	}

	/** @test */
	public function method_conditional() {
		$cbCallCount = 0;
		$isFooBar = function ($argA, $argB) {
			return $argA == 'foo' && $argB == 'bar';
		};
		$assertFooBar = function ($argA, $argB) use (&$cbCallCount) {
			$cbCallCount++;
			$this->assertEquals('foo', $argA);
			$this->assertEquals('bar', $argB);
		};

		$assertWhenFooBar = Fn\conditional($isFooBar, $assertFooBar);

		$assertWhenFooBar('faz', 'baz');
		$assertWhenFooBar('foo', 'shnoz');
		$this->assertEquals(0, $cbCallCount);

		$assertWhenFooBar('foo', 'bar');
		$this->assertEquals(1, $cbCallCount);
	}

	/** @test */
	public function method_concat() {
		$this->assertEquals(
			['a', 'b', 'c'],
			Fn\concat(['a', 'b'], 'c'),
			'Should add a value to an array'
		);
		$this->assertEquals(
			['a', 'b', 'c', 'd'],
			Fn\concat(['a', 'b'], ['c', 'd']),
			'Should merge an array'
		);
	}

	/** @test */
	public function method_pluck() {
		$this->assertEquals(
			['moe', 'larry', 'curly'],
			Fn\pluck([
				['name' => 'moe', 'age' => 45],
				['name' => 'larry', 'age' => 55],
				['name' => 'curly', 'age' => 65]
			], 'name'),
			'Should pluck item properties'
		);

		$this->assertEquals(
			[45, 55],
			Fn\pluck([
				['name' => 'moe', 'age' => 45],
				['name' => 'larry', 'age' => 55],
				['name' => 'curly']
			], 'age'),
			'Should handle missing properties'
		);

		$this->assertEquals(
			[],
			Fn\pluck([], 'name'),
			'Should handle empty collections'
		);
	}

	/** @test */
	public function method_times_shouldCallAFunctionXNumberOfTimes() {
		Fn\times(3, $spy = new Spy());

		$spy->shouldHaveBeenCalled()
			->times(3);
	}

	/** @test */
	public function method_times_shouldCallZeroTimes() {
		Fn\times(0, $spy = new Spy());

		$spy->shouldNotHaveBeenCalled();
	}

	/** @test */
	public function method_times_shouldCallWithIndex() {
		Fn\times(3, $spy = new Spy());

		$spy->shouldHaveBeenCalled()
			->with(0);
		$spy->shouldHaveBeenCalled()
			->with(1);
		$spy->shouldHaveBeenCalled()
			->with(2);
	}

	/**
	 * @test
	 * @expectedException \InvalidArgumentException
	 */
	public function method_times_shouldHandleNegativeArg() {
		Fn\times(-1, $spy = new Spy());

		$spy->shouldNotHaveBeenCalled();
	}
}


tests/Fixture/SomeClass.php
<?php


namespace Aeris\FnTest\Fixture;


class SomeClass {

	public $ctorArgs;

	public function __construct() {
		$this->ctorArgs = func_get_args();
	}

	public function getSomeProp() {
		return 'someVal';
	}

	public function passThru($arg) {
		return $arg;
	}

}

src/Fn.php
<?php
namespace Aeris\Fn;

function partial($func /*, $var_args*/) {
	$applied_args = array_slice($args = func_get_args(), 1);

	return function () use ($func, $applied_args) {
		$args = array_merge($applied_args, func_get_args());
		return call_user_func_array($func, $args);
	};
}

// adapted from https://gist.github.com/adaburrows/941874
function compose($fnA, $fnB) {
	return function () use ($fnA, $fnB) {
		$args = func_get_args();
		return $fnA(call_user_func_array($fnB, $args));
	};
}

function identity() {
	return function ($val) {
		return $val;
	};
}

function constant($val) {
	return function () use ($val) {
		return $val;
	};
}


function always() {
	return function () {
		return true;
	};
}

function never() {
	return function () {
		return false;
	};
}


function cat($item, array $array) {
	return array_merge(array($item), $array);
}

function moreThanOrEqualTo($min) {
	return function ($val) use ($min) {
		return $val >= $min;
	};
}

function lessThan($max) {
	return function ($val) use ($max) {
		return $val < $max;
	};
}

function equalTo($target) {
	return function ($val) use ($target) {
		return $val === $target;
	};
}

function notEqualTo($target) {
	$equalToTarget = partial(cb('equalTo'), $target);

	return negate($equalToTarget);
}

function even() {
	return function ($num) {
		return $num % 2 === 0;
	};
}

function odd() {
	return negate(even());
}

function both($predA, $predB) {
	return function ($val) use ($predA, $predB) {
		return resolve($predA, $val) && resolve($predB, $val);
	};
}

function invoker( /*$var_fns*/) {
	$fns = func_get_args();

	return function () use ($fns) {
		$args = func_get_args();

		foreach ($fns as $fn) {
			call_user_func_array($fn, $args);
		}
	};
}

function invokeAll( /*$var_fns*/) {
	$fns = func_get_args();
	// Returns last result
	return array_reduce($fns, function ($carry, $fn) {
		return invoke($fn);
	});
}

function invoke($fn) {
	$appliedArgs = array_slice($args = func_get_args(), 1);

	return call_user_func_array($fn, $appliedArgs);
}

function negate($fnOrVal) {
	if (is_callable($fnOrVal)) {
		return function () use ($fnOrVal) {
			return !call_user_func_array($fnOrVal, func_get_args());
		};
	}
	return !$fnOrVal;
}

function conditional($predicate, $cb) {
	return function () use ($predicate, $cb) {
		$appliedArgs = func_get_args();
		$resolveArgs = cat($predicate, $appliedArgs);
		$predicateVal = call_user_func_array(cb('resolve'), $resolveArgs);

		if ($predicateVal) {
			call_user_func_array($cb, $appliedArgs);
		}
	};
}

function doWhen($predicate, $fn) {
	if (resolve($predicate)) {
		return $fn();
	}
	return null;
}


function errorThrower($ErrorType, $opt_msg = null) {
	return partial(cb('throwError'), $ErrorType, $opt_msg);
}

function conditionalThrower($predicate, $ErrorType, $opt_msg = null) {
	return conditional($predicate, errorThrower($ErrorType, $opt_msg));
}

function throwWhen($predicate, $ErrorType, $opt_msg = null) {
	doWhen($predicate, errorThrower($ErrorType, $opt_msg));
}

/**
 * @param string $ErrorType
 * @param null   $opt_msg
 */
function throwError($ErrorType, $opt_msg = null) {
	$error = is_callable($ErrorType) ? $ErrorType() : new $ErrorType($opt_msg);

	throw $error;
}

function resolve($fnOrVal/* $var_args*/) {
	$applied_args = array_slice($args = func_get_args(), 1);

	return is_callable($fnOrVal) ? call_user_func_array($fnOrVal, $applied_args) : $fnOrVal;
}

function resolver(/*$var_args*/) {
	$args = func_get_args();
	return function ($predicate) use ($args) {
		$argsForResolve = cat($predicate, $args);

		return call_user_func_array(cb('resolve'), $argsForResolve);
	};
}

/**
 * Creates a function which calls the specified method
 * with the array item, only if the array has the item.
 *
 * eg.
 *
 *  $config = array('foo' => 'bar');
 *
 *  $callObjectWithConfig = F::whenSetInvoker($obj, $config);
 *
 *  $callObjectWithConfig('setFoo', 'foo');
 *  // calls $obj->setFoo('foo')
 *
 *  $callObjectWithConfig('setFaz', 'faz');
 *  // does nothing
 *
 * @param array  $arr
 * @param object $opt_obj Object to bind method names against.
 *                        If not set, returned fn will expect callables as arguments.
 * @return callable
 */
function whenSetInvoker(array $arr, $opt_obj = null) {
	return function ($method, $key) use ($opt_obj, $arr) {
		if (isKeySet($arr, $key)) {
			$cb = $opt_obj ? array($opt_obj, $method) : $method;
			call_user_func($cb, $arr[$key]);
		}
	};
}

function keySetChecker(array $arr) {
	return partial(cb('isKeySet'), $arr);
}

function isKeySet(array $arr, $key) {
	return isset($arr[$key]);
}

function accessor($key) {
	return function (array $arr) use ($key) {
		return $arr[$key];
	};
}

function cb($methodName) {
	return __NAMESPACE__ . '\\' . $methodName;
}


function zip($arr1, $arr2) {
	return array_map(null, $arr1, $arr2);
}


function mapAssoc(array $arr, $cb) {
	$mappedArr = array();

	foreach ($arr as $key => $val) {
		$mappedArr = $cb($val, $key);
	}

	return $mappedArr;
}


function pick(array $arr /*,$var_keys*/) {
	$pickedArr = array();
	$whitelistedKeys = array_slice($args = func_get_args(), 1);

	foreach ($whitelistedKeys as $key) {
		$pickedArr[$key] = $arr[$key];
	}

	return $pickedArr;
}

/**
 * Plucks a property from a collection of associative arrays.
 *
 * eg:
 * 	Fn\pluck([
 * 			['name' => 'moe', 'age' => 45],
 * 			['name' => 'larry', 'age' => 55],
 * 			['name' => 'curly', 'age' => 65]
 * 		], 'name')
 * // ['moe', 'larry', 'curly']
 *
 * @param array[] $collection
 * @param string $propName
 * @return mixed
 */
function pluck(array $collection, $propName) {
	return array_reduce($collection, function($vals, $item) use ($propName) {
		return isset($item[$propName]) ? concat($vals, $item[$propName]) : $vals;
	}, []);
}

/**
 * Add a value to an array, or merge a set of values.
 *
 * Fn\concat(['a', 'b'], 'c');  				// ['a', 'b', 'c']
 * Fn\concat(['a', 'b'], ['c', 'd']);  	// ['a', 'b', 'c', 'd']
 *
 * @param array $arr
 * @param array|mixed $valOrArray
 * @return array
 */
function concat(array $arr, $valOrArray) {
	$arrToMerge = is_array($valOrArray) ? $valOrArray : [$valOrArray];
	return array_merge([], $arr, $arrToMerge);
}

/**
 * Returns true if any item in the array
 * pass the $test callable.
 *
 * @param array    $arr
 * @param callable $test
 * @return boolean
 */
function any(array $arr, $test = null) {
	return array_reduce($arr, function ($anyIsTrue, $item) use ($test) {
		$itemIsTrue = $test ? call_user_func($test, $item) : $item;

		return $anyIsTrue || $itemIsTrue;
	}, false);
}

function all(array $arr, $test = null) {
	return array_reduce($arr, function ($anyIsTrue, $item) use ($test) {
		$itemIsTrue = $test ? call_user_func($test, $item) : $item;

		return $anyIsTrue && $itemIsTrue;
	}, true);
}

function joinOr(array $predicates) {
	return joinUsing($predicates, cb('any'));
}

function joinAnd(array $predicates) {
	return joinUsing($predicates, cb('all'));
}

function joinUsing(array $predicates, $mapTest) {
	return function (/** $var_args */) use ($predicates, $mapTest) {
		$args = func_get_args();

		// Call each filter with the applied args,
		// and check the results with
		// either 'any' or 'all'
		return call_user_func($mapTest, $predicates, function ($pred) use ($args) {
			return call_user_func_array($pred, $args);
		});
	};
}

/**
 * Return a factory for creating an object.
 *
 * eg.
 * $entityFactory = F::factory('MyApp\Model\Entity');
 * $entity = $entityFactory($ctorArg);  // => ~ new MapApp\Model\Entity($ctorArg)
 *
 * @param $className
 * @return callable
 */
function factoryMap($className) {
	return function ($data) use ($className) {
		return array_map(factory($className), $data);
	};
}

/**
 * Return a factory for creating an object.
 *
 * eg.
 * $entityFactory = F::factory('MyApp\Model\Entity');
 * $entity = $entityFactory($ctorArg);  // => ~ new MapApp\Model\Entity($ctorArg)
 *
 * @param string $className
 * @return callable
 */
function factory($className) {
	return function ($data) use ($className) {
		$ctorArgs = func_get_args();
		$reflector = new \ReflectionClass($className);
		return $reflector->newInstanceArgs($ctorArgs);
	};
}

function find(array $collection, callable $predicate) {
	foreach ($collection as $key => $val) {
		if ($predicate($val, $key)) {
			return $val;
		}
	}
	return null;
}

/**
 * eg.
 *  $obj->getFoo = function() { return 'foo' }
 *
 *  $getFooCaller = F::caller('getFoo');
 *  $getFooCaller($obj);  // => 'foo'
 *
 * Looks useless, but check this out:
 *
 *  array_map(F::caller('getFoo'), [$objA, $objB, $objC])  // => ['foo', 'foo', 'foo']
 *
 * @param string $methodName
 * @return callable
 */
function caller($methodName) {
	return function ($obj) use ($methodName) {
		$args = array_slice($args = func_get_args(), 1);
		$methodCallable = array($obj, $methodName);

		return call_user_func_array($methodCallable, $args);
	};
}

/**
 * @param array         $array
 * @param bool|callable $wherePredicate
 * @return number The count of items in the array which match the predicate.
 */
function countWhere(array $array, $wherePredicate = true) {
	return array_reduce($array, function ($sum, $item) use ($wherePredicate) {
		return $sum + (resolve($wherePredicate, $item) ? 1 : 0);
	}, 0);
}

/**
 * @param array $array
 * @param       $mapper
 */
function reduceAssoc(array $array, $mapper, $init) {
	$res = $init;

	foreach ($array as $key => $val) {
		$res = call_user_func($mapper, $res, $val, $key);
	};

	return $res;
}

function times($times, callable $cb) {
	if ($times < 0) {
		throw new \InvalidArgumentException('Fn\times arg must be at least 0');
	}

	$runCount = 0;

	while ($runCount < $times) {
		$cb($runCount);
		$runCount++;
	}
}